### PR TITLE
feat: new backup wallet

### PR DESCRIPTION
**Pull request by nuanyang233**
## Description

Closes #MF-4613

## Type of change

- [ ] Documentation
- [ ] Code refactoring (Restructuring existing code w/o changing its observable behavior)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (a fix or feature that would make something no longer possible to do/require old user must upgrade their Mask Network to this new version)

## Checklist

- [ ] My code follows the style guidelines of this project.
- [ ] I have performed a self-review of my own code.
  - [ ] I have removed all in development `console.log`s
  - [ ] I have removed all commented code.
- [ ] I have commented on my code, particularly in hard-to-understand areas.
- [ ] I have read [Internationalization Guide](https://github.com/DimensionDev/Maskbook/blob/develop/docs/i18n-guide.md) and moved text fields to the i18n JSON file.

If this PR depends on external APIs:

- [ ] I have configured those APIs with CORS headers to let extension requests get passed. 
  - chrome extension: `chrome-extension://[id]`
  - firefox extension: `moz-extension://[id]`
- [ ] I have delegated all web requests to the background service via the internal RPC bridge.

### CODE REVIEW

File: packages/shared-base/src/types/Wallet.ts [hunk 1]
```diff
@@ -42,6 +42,8 @@ export interface Wallet {
     deployed?: boolean
     /** persona identifier */
     identifier?: string
+    /** primary wallet address for derived wallets */
+    primaryWallet?: string
```
Comment (guanbinrui) [inline packages/shared-base/src/types/Wallet.ts]: Use `mnemonicId` as the id of the `hash(mnemonic)`.

File: packages/shared-base/src/types/Wallet.ts [hunk 2]
```diff
@@ -42,6 +42,8 @@ export interface Wallet {
     deployed?: boolean
     /** persona identifier */
     identifier?: string
+    /** primary wallet address for derived wallets */
```
Comment (guanbinrui) [inline packages/shared-base/src/types/Wallet.ts]:
```suggestion
/** The 'primaryAddress' represents the very first or genesis wallet address
  * derived from a particular mnemonic phrase. All subsequent wallets that 
  * are derived from the same mnemonic will reference this address, allowing 
  * for easy grouping and identification of wallets stemming from the same seed.
  */

```

File: packages/shared-base/src/types/Wallet.ts (hunk 1 above)
Comment (guanbinrui) [inline packages/shared-base/src/types/Wallet.ts]:
```suggestion
    primaryAddress?: string
```

Comment (guanbinrui) [reply packages/shared-base/src/types/Wallet.ts]: Or `genisisAddress`, `groupAddress`.

File: packages/mask/background/services/wallet/services/wallet/index.ts [hunk 3]
```diff
@@ -68,6 +77,18 @@ export async function getWalletPrimary() {
...
+export async function getMnemonicId(storedKeyInfo: api.IStoredKeyInfo) {
+    const masterPassword = await password.INTERNAL_getMasterPasswordRequired()
+    const firstWallet = await Mask.createAccountOfCoinAtPath({
+        coin: api.Coin.Ethereum,
+        password: masterPassword,
+        derivationPath: `${HD_PATH_WITHOUT_INDEX_ETHEREUM}/0`,
+        StoredKeyData: storedKeyInfo.data,
+    })
+    if (!firstWallet?.account?.address) return
+    return firstWallet?.account?.address
```
Comment (guanbinrui) [inline packages/mask/background/services/wallet/services/wallet/index.ts]: formatEthereumAddress

File: packages/mask/background/services/wallet/services/wallet/index.ts [hunk 4]
```diff
@@ -304,19 +347,25 @@ export function createWalletFromMnemonicWords(
...
     name: string,
     mnemonic: string,
     derivationPath = `${HD_PATH_WITHOUT_INDEX_ETHEREUM}/0`,
+    mnemonicId?: string,
 ) {
     return addWalletFromMnemonicWords(ImportSource.UserProvided, name, mnemonic, derivationPath)
 }
 
-export function recoverWalletFromPrivateKey(name: string, privateKey: string) {
+export function recoverWalletFromPrivateKey(name: string, privateKey: string, mnemonicId?: string) {
```
Comment (guanbinrui) [inline packages/mask/background/services/wallet/services/wallet/index.ts]: `mnemonicId` not used.

File: packages/mask/background/services/wallet/services/wallet/index.ts [hunk 5]
```diff
@@ -304,19 +347,25 @@ export function createWalletFromMnemonicWords(
...
+    derivationPath?: string,
+) {
+    return addWalletFromPrivateKey(ImportSource.LocalGenerated, name, privateKey, mnemonicId, derivationPath)
 }
 
 export function recoverWalletFromMnemonicWords(
     name: string,
     mnemonic: string,
     derivationPath = `${HD_PATH_WITHOUT_INDEX_ETHEREUM}/0`,
+    mnemonicId?: string,
```
Comment (guanbinrui) [inline packages/mask/background/services/wallet/services/wallet/index.ts]: `mnemonicId` not used.

File: packages/mask/background/services/wallet/services/wallet/index.ts (hunk 3 above)
Comment (nuanyang233) [reply packages/mask/background/services/wallet/services/wallet/index.ts]: This method is used to get the address as `mnemonicId`, sometimes you need to use `isSameAddress` to determine if it's the same address, so you don't need format.

File: packages/mask/background/services/wallet/services/wallet/database/wallet.ts [hunk 6]
```diff
@@ -43,6 +43,12 @@ export async function hasStoredKeyInfo(storedKeyInfo?: api.IStoredKeyInfo) {
     return wallets.filter((x) => x.storedKeyInfo?.hash).some((x) => x.storedKeyInfo?.hash === storedKeyInfo?.hash)
 }
 
+export async function getWalletByStoreKeyInfo(storedKeyInfo?: api.IStoredKeyInfo) {
```
Comment (guanbinrui) [inline packages/mask/background/services/wallet/services/wallet/database/wallet.ts]: Not used?

File: packages/mask/background/services/wallet/services/wallet/index.ts [hunk 7]
```diff
@@ -68,6 +77,28 @@ export async function getWalletPrimary() {
...
+        coin: api.Coin.Ethereum,
+        password: masterPassword,
+        derivationPath: `${HD_PATH_WITHOUT_INDEX_ETHEREUM}/0`,
+        StoredKeyData: storedKeyInfo.data,
+    })
+    if (!firstWallet?.account?.address) return
+    return firstWallet?.account?.address
+}
+
+export async function getMnemonicIdByWords(mnemonic: string) {
```
Comment (guanbinrui) [inline packages/mask/background/services/wallet/services/wallet/index.ts]: Too many query methods. Please store the `mnemonicId` in `sha3(words)`. It's computable everywhere and enough with the query method `getWalletByMnemonicId()`.

**Review by guanbinrui — APPROVED** [review]

**Pull request merged**